File: java/ZDD.java
import java.util.*;
import java.io.*;

public class ZDD {
    
    Hash hash;
    int osize;
    int numberOfNode;
    int root;
    int[] f;
    HashMap<int[],Integer> cache;
    HashMap<Integer,Integer> countCache;

    ZDD(ArrayList<String> rulelist,int hsize,int osize){
	hash = new Hash(hsize);
	osize = osize;
	numberOfNode = 0;
	f = new int[hsize];
	cache = new HashMap<int[],Integer>();
	for(int i = 0; i <= rulelist.size(); i++){
	    hash.insert(-1,i,-1,-1);
	    ++numberOfNode;
	}


	for(int i = 1; i <= rulelist.size(); i++){
	    int node = hash.member(-1,i,-1,-1);
	    // System.out.println(node);
	    for(int j = rulelist.get(0).length()-1; j > -1; --j) {
		if(rulelist.get(i-1).charAt(j) == '1') {
		    node = hash.insert(j+1,-1,0,node);
		    ++numberOfNode;
		}
		if(rulelist.get(i-1).charAt(j) == '*') {
		    node = hash.insert(j+1,-1,node,node);
		    ++numberOfNode;
		}
	    }		
	    f[i-1] = node;
	}

	int node = f[0];
	for (int i = 1; i < rulelist.size(); ++i)
	    node = unification(node,f[i]);
	root = node;

	    

    }
    
    
    int getNode(int var, int val, int left, int right){
	if ( 0 == right) { return left; }

	int P = hash.member(var, val, left, right);
	if (-1 != P) { return P; }
	P = hash.insert(var, val, left, right);
	++numberOfNode;
	return P;
    }
        
    int unification(int P, int Q){
	int R = 0;
	int[] o = {P,Q};

	// System.out.println(P + ", " + Q);
	if(0 == P) {
	    return Q;
	}
	if (0 == Q || P == Q) {
	    return P;
	}
	if (-1 == hash.topVar(P) && -1 == hash.topVar(Q)) {
	    if (P < Q) { return P; }
	    else { return Q; }
	}
	if( null != cache.get(o) ){
	    return cache.get(o);
	}
	if ((hash.topVar(P) < hash.topVar(Q) && hash.topVar(P) != -1) || hash.topVar(Q) == -1) {
	    // System.out.println(hash.topVar(P) + " (<) " + hash.topVar(Q));
	    R = getNode(hash.topVar(P),hash.topVal(P),unification(hash.getLeft(P),Q),hash.getRight(P) );	    
	}
	else if((hash.topVar(Q) < hash.topVar(P) && hash.topVar(Q) != -1) || hash.topVar(P) == -1) {
	    // System.out.println(hash.topVar(P) + " (>) " + hash.topVar(Q));
	    R = getNode(hash.topVar(Q),hash.topVal(Q),unification(P,hash.getLeft(Q)),hash.getRight(Q) );	    
	}
	else if( hash.topVar(P) == hash.topVar(Q) ){
	    // System.out.println(hash.topVar(P) + " (=) " + hash.topVar(Q));
	    R =  getNode(hash.topVar(P),hash.topVal(P),unification(hash.getLeft(P),hash.getLeft(Q)),unification(hash.getRight(P),hash.getRight(Q)) );
	}

	cache.put(o,R);
	if(cache.size() > osize)
	    cache.clear();

	return R;
    }

    int count(int n){
	
    	int m = countSub(n,root,hash.getTable());
    	countCache.clear();
	return m;
    }

    int countSub(int n, int P, Element[] tb){
	if(0 == P){ return 0; }
	else if(n == P){ return 1; }
	if(tb[P].getVal() != 1){ return 0;}
	if( null != countCache.get(P) ){return cache.get(P);}
	int sum  = countSub(n, hash.getLeft(P), tb) + countSub(n, hash.getRight(P), tb);
	countCache.put(P,sum);
	return sum;
    }
    
    
    void print(){
	HashSet<Integer> hs = new HashSet<Integer>();
	printSub(root,hs);
    }

    void printSub(int i,HashSet<Integer> hs){
	if (-1 == i) { return; }
	if (hs.contains(i)) { return; }
	hs.add(i);
        System.out.println(i + ", " + hash.topVar(i) + ", " + hash.topVal(i) + ", " + hash.getLeft(i) + ", " + hash.getRight(i));
	printSub(hash.getLeft(i), hs);
	printSub(hash.getRight(i), hs);
    }
	

    
    public static void main(String[] args){ 	  
	try{
	    ArrayList<String> ruleList = new ArrayList<String>();
	    File input = new File(args[0]);
	    BufferedReader br = new BufferedReader(new FileReader(input));
	    //File output = new File(args[1]);
	    //BufferedWriter bw = new BufferedWriter(new FileWriter(output));
	    String rule;
	    int ruleNum,bit;

	    
	    while((rule = br.readLine()) != null)
		ruleList.add(rule);	
	    
	    ZDD bdd = new ZDD(ruleList,Integer.parseInt(args[1]),Integer.parseInt(args[2]));
	    
	    bdd.print();
	    
	}catch(FileNotFoundException e) {
	    e.printStackTrace();
	}catch(IOException e) {
	    e.printStackTrace();
	}	
      }
    
}

File: c++/src/zdd.cpp
/* zdd.cpp */

#include <zdd.hpp>

ZDD::ZDD(std::vector<std::string>& rulelist, int hsize, int osize) {
  _n = rulelist[0].size();
  _hash = new Hash(hsize, _n);
  _osize = osize;
  std::vector<int> f;

  _hash->insert(_n+1,0,-1,-1,1);
  for (unsigned i = 1; i <= rulelist.size(); ++i) 
    _hash->insert(_n+1,i,-1,-1,1);
  
  for (unsigned i = 1; i <= rulelist.size(); ++i)
    f.push_back(makeZDDforRule(i,rulelist[i-1]));

  int node = f[0];

  std::cout << f.size() << std::endl;

  for (unsigned i = 1; i < f.size(); ++i) {
    node = unification(node,f[i]);
    printf("R[%d] is finished\n", i+1);
  }
  _root = node;
}

int ZDD::topVar(int i) { return _hash->topVar(i); }
int ZDD::topVal(int i) { return _hash->topVal(i); }
int ZDD::getLeft(int i) { return _hash->getLeft(i); }
int ZDD::getRight(int i) { return _hash->getRight(i); }
unsigned ZDD::getCounter(int i) { return _hash->getCounter(i); }
void ZDD::incCounter(int i) { _hash->incCounter(i); }
void ZDD::decCounter(int i) { _hash->decCounter(i); }

/* Warning!! getNode of BDDs is different from one of ZDDs. */
int ZDD::getNode(int var, int val, int left, int right) {
  if (0 == right) { return left; }

  int P = _hash->member(var, val, left, right);
  if (-1 != P) {
    incCounter(P);
    return P;
  }
  P = _hash->insert(var, val, left, right, 1);
  return P;
}

int ZDD::makeZDDforRule(int node, std::string& rule) {
  // std::cout << node << ", " << rule << std::endl;
  for (int j = rule.size()-1; j > -1; --j) {
    if (rule[j] == '1') {
      node = _hash->insert(j+1,-1,0,node,1);
      incCounter(0);
    }
    if (rule[j] == '*') { node = _hash->insert(j+1,-1,node,node,2);}
  }
  return node;
}

int ZDD::unification(int P, int Q) {
  int R = 0;

  if (0 == P) { // std::cout << "0 == P\n";
    return Q; }
  if (0 == Q || P == Q) { // std::cout << "0 == Q\n";
    return P; }
  if (P == _n+1 && Q == _n+1) {
    if (P < Q) { return P; }
    else { return Q; }
  }
  auto itr = _cache.find(std::pair<int,int>(P,Q));
  if (_cache.end() != itr) { return itr->second; }
  if (topVar(P) < topVar(Q)) {
    // node P is farther than node Q
    // printf("P < Q : P = %d, Q = %d\n", P, Q);
    R = getNode(topVar(P), topVal(P), unification(getLeft(P),Q), getRight(P));
  }
  else if (topVar(P) > topVar(Q)) {
    // printf("P > Q : P = %d, Q = %d\n", P, Q);
    R = getNode(topVar(Q), topVal(Q), unification(P,getLeft(Q)), getRight(Q));
  }
  else if (topVar(P) == topVar(Q)) {
    // printf("P == Q : P = %d, Q = %d\n", P, Q);
    R = getNode(topVar(P), topVal(P), unification(getLeft(P),getLeft(Q)), unification(getRight(P),getRight(Q)));
  }
  _cache[std::pair<int,int>(P,Q)] = R;
  if (_cache.size() > _osize) { _cache.clear(); }
  
  return R;
}

int ZDD::union_(int P, int Q) {
  if (0 == P) { return Q; }
  if (0 == Q || P == Q) { return P; }

  int R = 0;
  auto itr = _cache.find(std::pair<int,int>(P,Q));
  if (_cache.end() != itr) { return itr->second; }
  if (topVar(P) < topVar(Q)) {
    R = getNode(topVar(P), topVal(P), union_(getLeft(P),Q), getRight(P));
  } else if (topVar(P) > topVar(Q)) {
    // printf("P > Q : P = %d, Q = %d\n", P, Q);
    R = getNode(topVar(Q), topVal(Q), unification(P,getLeft(Q)), getRight(Q));
  } else if (topVar(P) == topVar(Q)) {
    R = getNode(topVar(P), topVal(P), union_(getLeft(P),getLeft(Q)), union_(getRight(P),getRight(Q)));
  }
  _cache[std::pair<int,int>(P,Q)] = R;
  if (_cache.size() > _osize) { _cache.clear(); }
  
  return R;
}

int ZDD::count(int n) {
  int m = countSub(n, _root, _hash->getTable());
  _countCache.clear();
  return m;
}

int ZDD::countSub(int n, int P, element* tb) {
  if (0 == P) { return 0; }
  else if (n == P) { return 1; }
  if (tb[P].getVal() != -1) { return 0; }
  auto itr = _countCache.find(P);
  if (_countCache.end() != itr) { return itr->second; }
  int sum = countSub(n, getLeft(P), tb) + countSub(n, getRight(P), tb);
  _countCache[P] = sum;
  return sum;
}

std::list<std::string> ZDD::getMatchHeaders(int n) {
  std::list<std::string> h = getMatchHeadersSub(n, _root, _hash->getTable());
  _scache.clear();
  return h;
}

std::list<std::string> ZDD::getMatchHeadersSub(int n, int P, element* tb) {
  // std::cout << "P = " << P << ", var = " << tb[P].getVar() << std::endl;

  std::list<std::string> headers;
  if (n == P) { 
    headers.push_back("");
    return headers; 
  }
  if (tb[P].getVar() == _n+1) { return headers; }

  auto itr = _scache.find(P);
  if (_scache.end() != itr) { return itr->second; }

  std::list<std::string> left = getMatchHeadersSub(n, getLeft(P), tb);
  std::list<std::string> right = getMatchHeadersSub(n, getRight(P), tb);

  std::string ladd = "";
  std::string radd = "";

  {
    int l = tb[getLeft(P)].getVar() - tb[P].getVar() - 1;
    int r = tb[getRight(P)].getVar() - tb[P].getVar() - 1;
    for (int i = 0; i < l; ++i) { ladd += "0"; }
    for (int i = 0; i < r; ++i) { radd += "0"; }
  }
  // std::cout << "P = " << P << ", var = " << tb[P].getVar() << ladd << ", " << radd << std::endl;

  std::list<std::string>::iterator it, end;
  it = left.begin(), end = left.end();
  while (it != end) { *it = "0" + ladd + *it, ++it; }

  it = right.begin(), end = right.end();
  while (it != end) { *it = "1" + radd + *it, ++it; }

  left.splice(left.end(), right);
  _scache[P] = left;

  return left;
}

void ZDD::print() {
  std::unordered_set<int>* us = new std::unordered_set<int>();
  printSub(_root, _hash->getTable(), us);
  delete us;
}

void ZDD::printSub(int i, element* tb, std::unordered_set<int>* us) {
  if (-1 == i) { return; }
  auto itr = us->find(i);
  if (us->end() != itr) { return; }
  us->insert(i);
  std::cout << "id[" << i << "] " << tb[i] << std::endl;
  printSub(tb[i].getLeft(), tb, us);
  printSub(tb[i].getRight(), tb, us);
}

unsigned ZDD::getNumberOfNodes() {
  std::unordered_set<int>* us = new std::unordered_set<int>();
  unsigned num = getNumberOfNodesSub(_root, us);
  delete us;
  return num;
}

unsigned ZDD::getNumberOfNodesSub(int P, std::unordered_set<int>* us) {
  if (-1 == P) { return 0; }
  auto itr = us->find(P);
  if (us->end() != itr) { return 0; }
  us->insert(P);
  
  return 1 + getNumberOfNodesSub(getLeft(P),us) + getNumberOfNodesSub(getRight(P),us);
}
